core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/NormalizeOrInExpressionRewriter.java

```diff
@@ -0,0 +1,128 @@
+import io.trino.sql.tree.InPredicate;
+import io.trino.sql.tree.LogicalExpression;
+
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.sql.ExpressionUtils.and;
+import static io.trino.sql.ExpressionUtils.or;
+import static io.trino.sql.tree.ComparisonExpression.Operator.EQUAL;
+import static io.trino.sql.tree.LogicalExpression.Operator.AND;
+
+public final class NormalizeOrInExpressionRewriter
+{
+    public static Expression normalizeOrInExpression(Expression expression)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(), expression);
+    }
+
+    private NormalizeOrInExpressionRewriter() {}
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        @Override
+        public Expression rewriteLogicalExpression(LogicalExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            List<Expression> terms = node.getTerms().stream()
+                    .map(expression -> treeRewriter.rewrite(expression, context))
+                    .collect(toImmutableList());
+
+            if (node.getOperator() == AND) {
+                return and(terms);
+            }
+
+            ImmutableList.Builder<InPredicate> inPredicateBuilder = ImmutableList.builder();
+            ImmutableSet.Builder<Expression> expressionToSkipBuilder = ImmutableSet.builder();
+            ImmutableList.Builder<Expression> othersExpressionBuilder = ImmutableList.builder();
+            groupComparisonAndInPredicate(terms).forEach((expression, values) -> {
+                if (values.size() > 1) {
+                    inPredicateBuilder.add(new InPredicate(expression, mergeToInListExpression(values)));
+                    expressionToSkipBuilder.add(expression);
+                }
+            });
+
+            Set<Expression> expressionToSkip = expressionToSkipBuilder.build();
+            for (Expression expression : terms) {
+                if (expression instanceof ComparisonExpression comparisonExpression && comparisonExpression.getOperator() == EQUAL) {
+                    if (!expressionToSkip.contains(comparisonExpression.getLeft())) {
+                        othersExpressionBuilder.add(expression);
+                    }
+                }
+                else if (expression instanceof InPredicate inPredicate && inPredicate.getValueList() instanceof InListExpression) {
+                    if (!expressionToSkip.contains(inPredicate.getValue())) {
+                        othersExpressionBuilder.add(expression);
+                    }
+                }
+                else {
+                    othersExpressionBuilder.add(expression);
+                }
+            }
+
+            return or(ImmutableList.<Expression>builder()
+                    .addAll(othersExpressionBuilder.build())
```

Comment:
Won't this introduce some regression like if we have ` colA = 1 or (some_complex_expression) or colA IN (2, 3)` then it would be rewritten to `(some_complex_expression) or colA IN (1, 2, 3)` - `colA = 1` would be a less expensive to compute when compared `colA = 1`. No changes requested just a thought.

Reply:
Good point! Not sure how the actual execution logic here, if the logic is evaluating expression one by until meet one fit the condition just like the java code logic then this maybe worth to adjust, maybe we could add `othersExpression` behind?
I would like to investigate this one maybe as follow up is fine, WDYT? If it is great if you can suggest where the execution logic code cc @kokosing